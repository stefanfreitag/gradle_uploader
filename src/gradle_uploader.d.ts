import { Schedule } from '@aws-cdk/aws-events';
import { Bucket } from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';
export interface SlackProperties {
    /** The Slack webhook used to send messages.
     * Details on setting up a webhook can be found at https://api.slack.com/messaging/webhooks
    */
    readonly webhook: string;
}
export interface MailProperties {
    readonly subscribers: Array<string>;
}
export interface GradleUploaderProps {
    readonly slackProperties?: SlackProperties;
    readonly mailProperties?: MailProperties;
    readonly whitelist: Array<string>;
    readonly schedule?: Schedule;
}
export declare class GradleUploader extends Construct {
    constructor(scope: Construct, id: string, uploaderProperties: GradleUploaderProps);
    private createTrigger;
    private createFunction;
    private createLambdaLayer;
    /**
     * Add mail subscriptions to the SNS topic.
     * @param subscriber The mail addresses to add to the SNS topic as subscriber.
     */
    private createTopic;
    createBucket(whitelist: Array<string>): Bucket;
}
