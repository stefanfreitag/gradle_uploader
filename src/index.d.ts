export * from './gradle_uploader';
