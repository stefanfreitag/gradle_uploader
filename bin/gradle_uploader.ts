#!/usr/bin/env node
import "source-map-support/register";
import * as cdk from "@aws-cdk/core";
import { GradleUploaderStack } from "../lib/gradle_uploader-stack";
import { Schedule } from "@aws-cdk/aws-events";

const app = new cdk.App();
new GradleUploaderStack(app, "GradleUploaderStack", {
  subscribers: ["stefan.freitag@udo.edu"],
  schedule: Schedule.cron({ minute: "0", hour: "0", day: "*"}),
  whitelist: ["87.122.220.125/32", "87.122.210.146/32"],
});
