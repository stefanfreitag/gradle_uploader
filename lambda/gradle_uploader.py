from collections.abc import Set
import json
import logging
import os
from logging import INFO, error, getLogger
from urllib.request import Request, urlopen

import boto3
import jsonpickle
import requests
from botocore.exceptions import ClientError
from requests.exceptions import HTTPError

from enum import Enum, auto, unique

logger = getLogger()
logger.setLevel(INFO)


S3_BUCKET = "<undefined>"

GRADLE_CURRENT_VERSION_METAINFO_URL = "https://services.gradle.org/versions/current"


@unique
class GradleDistribution(Enum):
    """
    The different types of Gradle distributions that can be downloaded.
    """

    BIN = (auto(), "Binary only")
    ALL = (auto(), "Binary, Sources, Docs")
    BOTH = (auto(), "Both distributions")


class MetaInfo:
    def __init__(self, dictionary: dict):
        setattr(self, "version", dictionary["version"])
        setattr(self, "download_url", dictionary["downloadUrl"])
        setattr(self, "checksum_url", dictionary["checksumUrl"])
        setattr(self, "wrapper_checksum_url", dictionary["wrapperChecksumUrl"])

    def __str__(self):
        return "MetaInformation: " + jsonpickle.dumps(self, indent=4)


def get_meta_info() -> MetaInfo:
    """
    Download the information about the current Gradle version from the Internet.
    """
    result = requests.get(GRADLE_CURRENT_VERSION_METAINFO_URL)
    json_data = json.loads(result.text)
    return MetaInfo(json_data)


def download_file(url: str):
    """
    Download a remote file to the tmp folder.
    :param url:  The URL of the remote file.
    :param target: The target (filename and path) on the local filesystem.
    
    """

    file_name = os.path.basename(url)
    target = "/tmp/" + file_name
    try:
        result = requests.get(url, timeout=5)
        logging.info("Writing " + url + " to " + target)
        open(target, "wb").write(result.content)
    except requests.exceptions.Timeout:
        print("Timeout")
    except requests.exceptions.TooManyRedirects:
        print("Too many redirects")
    except requests.exceptions.RequestException:
        print("Other error")


def upload_version(url: str, bucket_name: str) -> bool:
    """
    Upload a file to a S3 bucket. The file name is taken from the provided url.
    :param url:  The URL of the remote file.
    :param bucket_name: The name of the S3 bucket
    """

    session = boto3.Session()
    s3_client = session.client("s3")
    try:
        file_name = os.path.basename(url)
        print("Uploading " + file_name + " ...")
        s3_client.upload_file("/tmp/" + file_name, bucket_name, file_name)
        print("Upload done")
    except ClientError as e:
        print(e)
        return False
    return True


def check_existence(meta_info: MetaInfo, bucket_name: str) -> bool:
    session = boto3.Session()
    s3_client = session.client("s3")
    try:
        s3_client.head_bucket(Bucket=bucket_name)
    except ClientError as error:
        logging.error(error)
    try:
        logging.info("Checking for Gradle version in S3 bucket " + S3_BUCKET)
        print("Checking for Gradle version in S3 bucket " + S3_BUCKET)
        file_name = os.path.basename(meta_info.download_url)
        response = s3_client.head_object(Key=file_name, Bucket=S3_BUCKET)
        logging.info(response)
        return True
    except ClientError:
        logging.error("Not found")
        print("Not found")
        return False


def create_slack_message(meta_info: MetaInfo) -> str:
    """Slack message formatting"""

    header: str = """{"type": "header", "text": { "type": "plain_text",  "text": "Gradle Uploader" }}"""

    body: str = (
        f"{{"
        f"  'type': 'section',"
        f"  'text': {{"
        f"      'type': 'mrkdwn',"
        f"      'text': 'The release of a new Gradle version was detected.\n New release version: {meta_info.version}\n Download URL: {meta_info.download_url}\nChecksum URL: {meta_info.checksum_url}'"
        f"}}"
        f"}}"
    )

    basic_block: str = (f"""{{ "text": "A new Gradle version is available.", "blocks": [ {header}, {body} ] }}""")
    return basic_block


def send_slack_message(message: str, webhook_url: str):
    print(message)
    req = Request(webhook_url, message.encode("utf-8"))
    try:
        with urlopen(req) as res:
            res.read()
            print(res)
            logger.info("Message posted.")
    except HTTPError as err:
        print(err)
        logger.error("Request failed: %d %s", err.code, err.reason)


def distribution_to_urls(meta_info: MetaInfo, distribution: str) -> Set:
    urls = set()
    if (distribution == GradleDistribution.BIN.name) or (
        distribution == GradleDistribution.BOTH.name
    ):
        urls.add(meta_info.download_url)
    if (distribution == GradleDistribution.ALL.name) or (
        distribution == GradleDistribution.BOTH.name
    ):
        url = meta_info.download_url.replace("-bin.zip", "-all.zip")
        urls.add(url)
    return urls


def main(event, context):
    global S3_BUCKET
    try:
        S3_BUCKET = os.environ["BUCKET_NAME"]
        TOPIC_ARN = os.environ["TOPIC_ARN"]
        GRADLE_DISTRIBUTION = os.environ["GRADLE_DISTRIBUTION"]

        logging.info("S3 Bucket used as target: " + S3_BUCKET)
        meta_info = get_meta_info()

        logging.info("Current Gradle Version:" + meta_info.version)

        # Check based on distribution type
        if check_existence(meta_info, S3_BUCKET):
            logging.info(
                "No action required. Latest Gradle version "
                + meta_info.version
                + " available in S3."
            )
        else:
            download_urls = distribution_to_urls(
                meta_info=meta_info, distribution=GRADLE_DISTRIBUTION
            )
            for entry in download_urls:
                download_file(entry)
                upload_version(entry, bucket_name=S3_BUCKET)
            ##Send mail
            sns = boto3.client("sns")
            response = (
                sns.publish(
                    TopicArn=TOPIC_ARN,
                    Message="New Gradle version available: " + str(meta_info.version),
                ),
            )
            print("Sent mail")

            slack_message = create_slack_message(meta_info)
            send_slack_message(slack_message, os.environ["WEBHOOK_URL"])

    except KeyError as identifier:
        print("Key error for " + str(identifier))


if __name__ == "__main__":
    logging.basicConfig(filename="gradleUploader.log", level=logging.INFO)
    main(None, None)
