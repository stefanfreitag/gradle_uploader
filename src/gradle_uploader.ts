import * as path from 'path';
import {
  Function,
  Runtime,
  LayerVersion,
  Code,
} from 'aws-cdk-lib//aws-lambda';
import { RetentionDays } from 'aws-cdk-lib//aws-logs';
import { Topic } from 'aws-cdk-lib//aws-sns';
import { EmailSubscription } from 'aws-cdk-lib//aws-sns-subscriptions';
import { Rule, Schedule } from 'aws-cdk-lib/aws-events';
import { LambdaFunction } from 'aws-cdk-lib/aws-events-targets';
import { AnyPrincipal, Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import {
  Bucket,
  BlockPublicAccess,
  BucketEncryption,
  BucketPolicy,
} from 'aws-cdk-lib/aws-s3';
import { CfnOutput, Duration, RemovalPolicy } from 'aws-cdk-lib/core';
import { Construct } from 'constructs';

/**
 * Properties related to forwarding messages to Slack.
 */
export interface SlackProperties{
  /** The Slack webhook used to send messages. Details on setting up a webhook can be found at https://api.slack.com/messaging/webhooks.
   *  @experimental
  */
  readonly webhook: string;
}

/**
 * Properties related to forwarding messages via mail.
 */
export interface MailProperties {

  readonly subscribers: Array<string>;
}

/**
 * Types of available Gradle distributions.
 */
export enum GradleDistribution{
  /**
   * Binaries only.
   */
  BIN,
  /**
   * Binaries, sources and documentation.
   */
  ALL,
  /**
   * BINARY and ALL.
   */
  BOTH
}

export interface UploaderProperties {
  /**
   * Optional properties required for sending messages via Slack.
   */
  readonly slackProperties?: SlackProperties;
  /**
   * Optional properties required for sending messages via mail.
   */
  readonly mailProperties?: MailProperties;
  readonly whitelist: Array<string>;
  readonly schedule?: Schedule;
  /**
   * The {@link GradleDistribution | Gradle distribution} type to download. If no value is specified, only the binaries will be downloaded.
   */
  readonly distribution?: GradleDistribution;
}

export class GradleUploader extends Construct {
  constructor(
    scope: Construct,
    id: string,
    uploaderProperties: UploaderProperties ) {
    super(scope, id);

    const topic = this.createTopic();
    if (uploaderProperties.mailProperties?.subscribers) {
      this.addSubscribers(topic, uploaderProperties.mailProperties?.subscribers );
    }
    const bucket = this.createBucket(uploaderProperties.whitelist);
    const layer = this.createLambdaLayer();
    const fn = this.createFunction(layer, bucket, topic);

    if (uploaderProperties.slackProperties?.webhook) {
      fn.addEnvironment('WEBHOOK_URL', uploaderProperties.slackProperties?.webhook);
    }

    let distribution: GradleDistribution;
    if (uploaderProperties.distribution === undefined) {
      distribution = GradleDistribution.BIN;
    } else {
      distribution = uploaderProperties.distribution;
    }
    fn.addEnvironment('GRADLE_DISTRIBUTION', GradleDistribution[distribution]);


    bucket.grantReadWrite(fn);
    topic.grantPublish(fn);

    const schedule =
      uploaderProperties.schedule != null
        ? uploaderProperties.schedule
        : Schedule.cron({ minute: '0', hour: '0', day: '*' });
    this.createTrigger(fn, schedule);

    new CfnOutput(this, 'bucketArnOutput', {
      value: bucket.bucketArn,
      description: 'Bucket ARN',
    });
  }

  private createTrigger(fn: Function, schedule: Schedule) {
    const target = new LambdaFunction(fn);
    new Rule(this, 'ScheduleRule', {
      schedule: schedule,
      targets: [target],
    });
  }

  private createFunction(layer: LayerVersion, bucket: Bucket, topic: Topic) {

    return new Function(this, 'fnUpload', {
      runtime: Runtime.PYTHON_3_8,
      description: 'Download Gradle distribution to S3 bucket',
      code: Code.fromAsset(path.join(__dirname, '../lambda')),
      handler: 'gradle_uploader.main',
      timeout: Duration.minutes(5),
      memorySize: 512,
      logRetention: RetentionDays.ONE_WEEK,
      layers: [layer],
      environment: {
        BUCKET_NAME: bucket.bucketName,
        TOPIC_ARN: topic.topicArn,
      },
    });
  }

  private createLambdaLayer(): LayerVersion {
    return new LayerVersion(this, 'GradleUploaderLayer', {
      code: Code.fromAsset(path.join(__dirname, '../layer-code')),
      compatibleRuntimes: [Runtime.PYTHON_3_8],
      license: 'Apache-2.0',
      description: 'A layer containing dependencies for the Gradle Uploader',
    });
  }

  /**
   * Add mail subscriptions to the SNS topic.
   * @param subscriber The mail addresses to add to the SNS topic as subscriber.
   */
  private addSubscribers(topic: Topic, subscribers:Array<string>) {
    for (var subscriber of subscribers) {
      topic.addSubscription(new EmailSubscription(subscriber));
    }
  }
  private createTopic() {
    return new Topic(this, 'NotificationTopic', {});
  }

  createBucket(whitelist: Array<string>): Bucket {
    const bucket = new Bucket(this, 'bucket', {
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      publicReadAccess: false,
      versioned: false,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const bucketContentStatement = new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['s3:GetObject'],
      resources: [bucket.bucketArn + '/*'],
      principals: [new AnyPrincipal()],
      conditions: {
        IpAddress: {
          'aws:SourceIp': whitelist,
        },
      },
    });

    const bucketStatement: PolicyStatement = new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['s3:ListBucket', 's3:GetBucketLocation'],
      resources: [bucket.bucketArn],
      principals: [new AnyPrincipal()],
      conditions: {
        IpAddress: {
          'aws:SourceIp': whitelist,
        },
      },
    });

    const bucketPolicy = new BucketPolicy(this, 'bucketPolicy', {
      bucket: bucket,
    });

    bucketPolicy.document.addStatements(
      bucketContentStatement,
      bucketStatement,
    );
    return bucket;
  }
}
