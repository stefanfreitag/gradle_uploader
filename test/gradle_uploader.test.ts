import {
  expect as expectCDK,
  haveResource,
  haveResourceLike,
} from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { GradleDistribution, GradleUploader } from '../src/gradle_uploader';

test('S3 bucket is encrypted and not public accessible ', () => {
  // Given
  const stack = new Stack();
  new GradleUploader(stack, 'MyConstruct', {
    mailProperties: { subscribers: ['john.doe@foobar.com'] },
    whitelist: ['87.122.220.125/32', '87.122.210.146/32'],
  });

  // Then
  expectCDK(stack).to(
    haveResource('AWS::S3::Bucket', {
      PublicAccessBlockConfiguration: {
        BlockPublicAcls: true,
        BlockPublicPolicy: true,
        IgnorePublicAcls: true,
        RestrictPublicBuckets: true,
      },
      BucketEncryption: {
        ServerSideEncryptionConfiguration: [
          {
            ServerSideEncryptionByDefault: {
              SSEAlgorithm: 'AES256',
            },
          },
        ],
      },
    }),
  );
});

test('SNS topic is setup ', () => {
  const stack = new Stack();

  new GradleUploader(stack, 'MyConstruct', {
    mailProperties: { subscribers: ['john.doe@foobar.com'] },
    whitelist: ['87.122.220.125/32', '87.122.210.146/32'],
  });

  expectCDK(stack).to(haveResource('AWS::SNS::Topic'));
});

test('SNS subscription is setup ', () => {
  const stack = new Stack();

  new GradleUploader(stack, 'MyTestStack', {
    mailProperties: { subscribers: ['john.doe@foobar.com'] },
    whitelist: ['87.122.220.125/32', '87.122.210.146/32'],
  });

  expectCDK(stack).to(
    haveResource('AWS::SNS::Subscription', {
      Protocol: 'email',
      Endpoint: 'john.doe@foobar.com',
    }),
  );
});

test('Lambda function is setup ', () => {
  const stack = new Stack();

  new GradleUploader(stack, 'MyTestStack', {
    mailProperties: { subscribers: ['john.doe@foobar.com'] },
    whitelist: ['87.122.220.125/32', '87.122.210.146/32'],
  });

  expectCDK(stack).to(
    haveResource('AWS::Lambda::Function', {
      Runtime: 'python3.8',
    }),
  );
});

test('Slack Webhook defined as Lambda environment variable', () => {
  const stack = new Stack();

  new GradleUploader(stack, 'MyConstruct', {
    mailProperties: { subscribers: ['john.doe@foobar.com'] },
    whitelist: ['87.122.220.125/32', '87.122.210.146/32'],
    slackProperties: {
      webhook:
        'https://hooks.slack.com/services/T00000000/B00000000/XXXXXXXXXXXXXXXXXXXXXXXX',
    },
  });

  expectCDK(stack).to(
    haveResourceLike('AWS::Lambda::Function', {
      Environment: {
        Variables: {
          WEBHOOK_URL:
            'https://hooks.slack.com/services/T00000000/B00000000/XXXXXXXXXXXXXXXXXXXXXXXX',
          GRADLE_DISTRIBUTION: 'BIN',
        },
      },
    }),
  );
  expectCDK(stack).to(
    haveResourceLike('AWS::SNS::Subscription', {
      Endpoint: 'john.doe@foobar.com',
      Protocol: 'email',
    }),
  );
});

test('Gradle Distribution selection forwarded to Lambda', () => {
  const stack = new Stack();

  new GradleUploader(stack, 'MyConstruct', {
    mailProperties: { subscribers: ['john.doe@foobar.com'] },
    whitelist: ['87.122.220.125/32', '87.122.210.146/32'],
    distribution: GradleDistribution.ALL,
  });

  expectCDK(stack).to(
    haveResourceLike('AWS::Lambda::Function', {
      Environment: {
        Variables: {
          GRADLE_DISTRIBUTION: 'ALL',
        },
      },
    }),
  );
});
