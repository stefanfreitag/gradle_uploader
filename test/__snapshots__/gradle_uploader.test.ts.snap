// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S3 bucket is encrypted and not public accessible  1`] = `
Object {
  "Outputs": Object {
    "MyConstructbucketArnOutputE285C4B4": Object {
      "Description": "Bucket ARN",
      "Value": Object {
        "Fn::GetAtt": Array [
          "MyConstructbucketC206B943",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParameters7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07ArtifactHashC98D9704": Object {
      "Description": "Artifact hash for asset \\"7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07\\"",
      "Type": "String",
    },
    "AssetParameters7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07S3BucketCFFE9496": Object {
      "Description": "S3 bucket for asset \\"7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07\\"",
      "Type": "String",
    },
    "AssetParameters7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07S3VersionKey80BB0DFB": Object {
      "Description": "S3 key for asset version \\"7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07\\"",
      "Type": "String",
    },
    "AssetParametersda0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505caArtifactHash69056A8A": Object {
      "Description": "Artifact hash for asset \\"da0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505ca\\"",
      "Type": "String",
    },
    "AssetParametersda0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505caS3BucketE1D97DF5": Object {
      "Description": "S3 bucket for asset \\"da0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505ca\\"",
      "Type": "String",
    },
    "AssetParametersda0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505caS3VersionKeyB49F21A7": Object {
      "Description": "S3 key for asset version \\"da0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505ca\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyConstructGradleUploaderLayerA35DA212": Object {
      "Properties": Object {
        "CompatibleRuntimes": Array [
          "python3.8",
        ],
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersda0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505caS3BucketE1D97DF5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersda0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505caS3VersionKeyB49F21A7",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersda0d323a0659d680b99f0c59aa2b64123262254858e5d3c90cce5e0c5e5505caS3VersionKeyB49F21A7",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "A layer containing dependencies for the Gradle Uploader",
        "LicenseInfo": "Apache-2.0",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "MyConstructNotificationTopic44794FFD": Object {
      "Type": "AWS::SNS::Topic",
    },
    "MyConstructNotificationTopicjohndoefoobarcomA5E4354F": Object {
      "Properties": Object {
        "Endpoint": "john.doe@foobar.com",
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "MyConstructNotificationTopic44794FFD",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "MyConstructScheduleRule997972F2": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(0 0 * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "MyConstructfnUpload8552058A",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "MyConstructScheduleRuleAllowEventRuleMyConstructScheduleRuleDDCD4BF4790CB693": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "MyConstructfnUpload8552058A",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "MyConstructScheduleRule997972F2",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MyConstructbucketC206B943": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "MyConstructbucketPolicy2322E323": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "MyConstructbucketC206B943",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Condition": Object {
                "IpAddress": Object {
                  "aws:SourceIp": Array [
                    "87.122.220.125/32",
                    "87.122.210.146/32",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": "*",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "MyConstructbucketC206B943",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:ListBucket",
                "s3:GetBucketLocation",
              ],
              "Condition": Object {
                "IpAddress": Object {
                  "aws:SourceIp": Array [
                    "87.122.220.125/32",
                    "87.122.210.146/32",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": "*",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "MyConstructbucketC206B943",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "MyConstructfnUpload8552058A": Object {
      "DependsOn": Array [
        "MyConstructfnUploadServiceRoleDefaultPolicy297AEF6E",
        "MyConstructfnUploadServiceRole51911E02",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07S3BucketCFFE9496",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07S3VersionKey80BB0DFB",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters7ab04dd4f1736f40464b2ed39139c1b346656f359bb91e287c5cffd798785e07S3VersionKey80BB0DFB",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "Download Gradle distribution to S3 bucket",
        "Environment": Object {
          "Variables": Object {
            "BUCKET_NAME": Object {
              "Ref": "MyConstructbucketC206B943",
            },
            "TOPIC_ARN": Object {
              "Ref": "MyConstructNotificationTopic44794FFD",
            },
          },
        },
        "Handler": "main",
        "Layers": Array [
          Object {
            "Ref": "MyConstructGradleUploaderLayerA35DA212",
          },
        ],
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyConstructfnUploadServiceRole51911E02",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyConstructfnUploadServiceRole51911E02": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyConstructfnUploadServiceRoleDefaultPolicy297AEF6E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "MyConstructbucketC206B943",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "MyConstructbucketC206B943",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "MyConstructNotificationTopic44794FFD",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyConstructfnUploadServiceRoleDefaultPolicy297AEF6E",
        "Roles": Array [
          Object {
            "Ref": "MyConstructfnUploadServiceRole51911E02",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
